File: leetcode_python/Depth-First-Search/weighted-sum-of-a-tree.py
"""
# https://leetcode.com/problems/weighted-sum-of-a-tree/description/


4015. Weighted Sum of a Tree
Solved
Medium
premium lock icon
Companies
Hint
You are given an integer array parent of length n representing a rooted tree with nodes labeled from 0 to n - 1.

The tree is rooted at node 0, so parent[0] = -1. For each node i where 1 <= i <= n - 1, parent[i] denotes the parent of node i.

You are also given an integer array nums of length n, where nums[i] denotes the value of node i.

The weight of a node i at depth d is nums[i] * (h - d + 1), where h is the height of the tree.

Return the sum of the weights of all nodes in the tree.

The depth of a node is the number of nodes on the path from the root to that node, inclusive, with the root having depth 1.

The height of the tree is the maximum depth among all nodes in the tree.

 

Example 1:

​​​​​​​

Input: parent = [-1,0,0,0,2,2], nums = [5,2,3,1,4,6]

Output: 37

Explanation:

The height of the tree is 3.

Node	nums[i]	Depth (d)	Weight
0	5	1	5 * (3 - 1 + 1) = 15
1	2	2	2 * (3 - 2 + 1) = 4
2	3	2	3 * (3 - 2 + 1) = 6
3	1	2	1 * (3 - 2 + 1) = 2
4	4	3	4 * (3 - 3 + 1) = 4
5	6	3	6 * (3 - 3 + 1) = 6
The sum of all node weights is 15 + 4 + 6 + 2 + 4 + 6 = 37.

Example 2:

​​​​​​​​​​​​​​

Input: parent = [-1,0,1,2], nums = [1,2,3,4]

Output: 20

Explanation:

The height of the tree is 4.

Node	nums[i]	Depth (d)	Weight
0	1	1	1 * (4 - 1 + 1) = 4
1	2	2	2 * (4 - 2 + 1) = 6
2	3	3	3 * (4 - 3 + 1) = 6
3	4	4	4 * (4 - 4 + 1) = 4
The sum of all node weights is 4 + 6 + 6 + 4 = 20.

 

Constraints:

1 <= n <= 105
n == parent.length == nums.length
parent[0] == -1
0 <= parent[i] <= n - 1 for all i in [1, n - 1]
1 <= nums[i] <= 106
The input is generated such that the array parent represents a valid tree rooted at node 0.


"""

# V0
class Solution(object):
    def weightedSum(self, parent, nums):
        """
        :type parent: List[int]
        :type nums: List[int]
        :rtype: int
        """
        pass


# V1
# IDEA: DFS (gpt)
class Solution(object):
    def weightedSum(self, parent, nums):
        """
        :type parent: List[int]
        :type nums: List[int]
        :rtype: int
        """
        n = len(parent)

        # depth[i] = depth of node i
        # root has depth 1
        self.depth = [0] * n
        self.parent = parent

        # Find depth of every node
        for i in range(n):
            self.helper(i)

        # Height = maximum depth
        height = max(self.depth)

        # Calculate weighted sum
        res = 0

        for i in range(n):
            res += nums[i] * (height - self.depth[i] + 1)

        return res

    def helper(self, x):
        """
        Return the depth of node x.
        """
        # Already calculated
        if self.depth[x] != 0:
            return self.depth[x]

        # Root
        if self.parent[x] == -1:
            self.depth[x] = 1
            return 1

        # Depth = parent's depth + 1
        self.depth[x] = self.helper(self.parent[x]) + 1

        return self.depth[x]

# V2

File: leetcode_java/src/main/java/LeetCodeJava/DFS/WeightedSumOfATree.java
package LeetCodeJava.DFS;

// https://leetcode.com/problems/weighted-sum-of-a-tree/

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.List;

/**
 *  4015. Weighted Sum of a Tree
 *  Medium
 *
 *  You are given an integer array parent of length n representing a rooted tree with
 *  nodes labeled from 0 to n - 1. The tree is rooted at node 0, so parent[0] = -1.
 *  For each node i where 1 <= i <= n - 1, parent[i] denotes the parent of node i.
 *
 *  You are also given an integer array nums of length n, where nums[i] is the value of
 *  node i.
 *
 *  The weight of a node i at depth d is nums[i] * (h - d + 1), where h is the height of
 *  the tree. The depth of a node is the number of nodes on the path from the root to
 *  that node inclusive (root has depth 1); the height is the maximum depth.
 *
 *  Return the sum of the weights of all nodes in the tree.
 *
 *  Example 1:
 *    Input: parent = [-1,0,0,0,2,2], nums = [5,2,3,1,4,6]
 *    Output: 37
 *
 *  Example 2:
 *    Input: parent = [-1,0,1,2], nums = [1,2,3,4]
 *    Output: 20
 *
 *  Constraints:
 *    1 <= n <= 10^5
 *    n == parent.length == nums.length
 *    parent[0] == -1
 *    0 <= parent[i] <= n - 1 for all i in [1, n - 1]
 *    1 <= nums[i] <= 10^6
 */
public class WeightedSumOfATree {

    // V0
    // IDEA: memoized depth walk up the parent chain (no children lists needed):
    //       depth(i) = depth(parent[i]) + 1, height = max depth, then sum the weights.
    //       Each node's depth is resolved once, so the total work stays linear.
    /**
     * time = O(n)
     * space = O(n)
     */
    public long weightedSum(int[] parent, int[] nums) {

        int n = parent.length;
        int[] depth = new int[n]; // 0 = not computed yet
        depth[0] = 1;

        int height = 1;
        int[] chain = new int[n]; // reusable buffer of unresolved ancestors

        for (int i = 0; i < n; i++) {
            if (depth[i] != 0) {
                height = Math.max(height, depth[i]);
                continue;
            }

            // walk up until a node with a known depth
            int len = 0;
            int cur = i;
            while (depth[cur] == 0) {
                chain[len++] = cur;
                cur = parent[cur];
            }

            int d = depth[cur];
            // fill the chain back down
            for (int k = len - 1; k >= 0; k--) {
                d++;
                depth[chain[k]] = d;
            }
            height = Math.max(height, d);
        }

        long res = 0L;
        for (int i = 0; i < n; i++) {
            res += (long) nums[i] * (height - depth[i] + 1);
        }

        return res;
    }

    // V1
    // IDEA: BFS LEVEL ORDER on an explicit children adjacency list rebuilt from `parent`.
    //       Depth comes from the BFS level, height is the deepest level reached.
    /**
     * time = O(n)
     * space = O(n)
     */
    public long weightedSum_1(int[] parent, int[] nums) {

        int n = parent.length;
        List<List<Integer>> children = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            children.add(new ArrayList<>());
        }
        for (int i = 1; i < n; i++) {
            children.get(parent[i]).add(i);
        }

        int[] depth = new int[n];
        depth[0] = 1;
        int height = 1;

        Deque<Integer> q = new ArrayDeque<>();
        q.add(0);
        while (!q.isEmpty()) {
            int cur = q.poll();
            height = Math.max(height, depth[cur]);
            for (Integer child : children.get(cur)) {
                depth[child] = depth[cur] + 1;
                q.add(child);
            }
        }

        long res = 0L;
        for (int i = 0; i < n; i++) {
            res += (long) nums[i] * (height - depth[i] + 1);
        }
        return res;
    }

    // V2
    // IDEA: brute force O(n * h) - kept as a readable correctness reference.
    //       For every node, walk the parent chain all the way up to the root and count the
    //       steps. No caching at all, so a long chain re-walks the same ancestors.
    /**
     * time = O(n * h), h = tree height (O(n^2) on a path-shaped tree)
     * space = O(n)
     */
    public long weightedSum_2(int[] parent, int[] nums) {

        int n = parent.length;
        int[] depth = new int[n];
        int height = 0;

        for (int i = 0; i < n; i++) {
            int d = 1;
            int cur = i;
            while (parent[cur] != -1) {
                cur = parent[cur];
                d++;
            }
            depth[i] = d;
            height = Math.max(height, d);
        }

        long res = 0L;
        for (int i = 0; i < n; i++) {
            res += (long) nums[i] * (height - depth[i] + 1);
        }
        return res;
    }
}
